### PR TITLE
Fix #4783 with a designated option

### LINKED ISSUE #4783: Inconsistent semantics after setDT
```
> d1 <- data.frame(a=c(1,2,3,4), b=c(2,3,4,5))
> d2 <- d1
> setDT(d2)  # At this point d2 is a shallow copy of d1, pointing to the same columns
```

Do modifications to `d2` impact `d1`?  We could live with both 'yes' or 'no', but the answer is _sometimes_:
```
d2[, b:=3:7]         # (1) impacts only d2
d2[, c:=4:8]         # (2) impacts only d2
d2[!is.na(a), b:=31] # (3) impacts both
d2[, b:=30]          # (4) impacts both
```
In cases 1&2 `d2` 'plunks' the full columns into itself and `d1` isn't affected. In cases 3 & 4 it seems that operation-in-place optimization kicks in (`address(d2$b)` is unchanged), so there is no copy-on-write and data still pointed to by `d1` is overwritten.

These semantic discrepancies make (the otherwise great) setDT unusable to us except in the most trivial scripts.

`#` `Output of sessionInfo()`
R version 3.6.1 (2019-07-05)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19041)
Matrix products: default

### PR DESCRIPTION
**Pull request by OfekShilon**
This PR adds an option that disables in-place assignment - thereby solving #4783 and also indirectly the old #3215.
The option is currently initialized to TRUE (old behaviour) - but at least for us the right initialization is FALSE (fixed behaviour).

Measurements on various size DTs show that the added performance overhead (extra allocations) is definitely manageable.

### CODE REVIEW

**Comment by OfekShilon** [comment]
This continues the discussion in [this PR](https://github.com/Rdatatable/data.table/pull/4902)

**Comment by OfekShilon** [comment]
I'm not sure how to continue - all tests pass for me locally.  Could it be a merge conflict I missed?

File: inst/tests/tests.Rraw [hunk 1]
```diff
@@ -17264,12 +17271,40 @@ if (identical(x, enc2native(x))) {
...
+
+# In-place assignment can leak to origin DF, #4783
+ orig.opt <- options(datatable.assign.inplace=TRUE)
+ d <- data.frame(a=c(1,2), b=c(1,2))
+ df <- d
+ setDT(d)
+ d[, a:=c(10, 11)]  # not leaking to df
+ test(2164.2, df$a[1], 1)
+ d[!is.na(a), b:=c(20,21)]  # leaks to df
+ test(2166, df$b[1], 20)
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:17291]: the issue is test numbering. here is 2166, next is 2164.3 -- tests should be in increasing order in the file.

Comment (OfekShilon) [reply inst/tests/tests.Rraw]: Thanks! I just added another commit (with other improvements) and hopefully this is fixed.

**Pull request closed without merging**